### ISSUE TITLE
Incorrect suggestion for patterns on non-exhaustive enums

### ISSUE THREAD

**Issue by ohsayan** [rust-lang/rust#85620]
## `if let`

Given the following code:

```rust
#[non_exhaustive]
enum Endless {
    JustOnce,
}
fn main() {
    let edl = Endless::JustOnce;
    if let Endless::JustOnce = edl {
        println!("Oooh");
    } else {
        println!("Ugh");
    }
}
```

The current output is: (https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=037b50b76d3a47d064ec9830a9f2d234)

```
warning: irrefutable `if let` pattern
  --> src/main.rs:7:5
   |
7  | /     if let Endless::JustOnce = edl {
8  | |         println!("Oooh");
9  | |     } else {
10 | |         println!("Ugh");
11 | |     }
   | |_____^
   |
   = note: `#[warn(irrefutable_let_patterns)]` on by default
   = note: this pattern will always match, so the `if let` is useless
   = help: consider replacing the `if let` with a `let`

warning: 1 warning emitted
```
**Ideally the output should look like:** No warnings reported because the enum is `#[non_exhaustive]`
## `match`

Given:
```rust
#[non_exhaustive]
enum Endless {
    JustOnce,
}
fn main() {
    let edl = Endless::JustOnce;
    match edl {
        Endless::JustOnce => {},
        wildcard => {
            // in the future, something will happen here
        }
    }
}
```

The current output is: (https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=27584713d984504b179ae80785e089ab)
```
warning: unreachable pattern
 --> src/main.rs:9:9
  |
9 |         wildcard => {
  |         ^^^^^^^^
  |
  = note: `#[warn(unreachable_patterns)]` on by default

warning: unused variable: `wildcard`
 --> src/main.rs:9:9
  |
9 |         wildcard => {
  |         ^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_wildcard`
  |
  = note: `#[warn(unused_variables)]` on by default

warning: 2 warnings emitted
```
**Ideally the output should look like:** Only the warning for the unused `wildcard` variable

**Comment by ohsayan** [comment]
Maybe related: https://github.com/camelid/rust/blob/5d2a2a1caa1a12a5c445fcc99e2dee42e5a5f3d7/compiler/rustc_mir_build/src/thir/pattern/check_match.rs#L368..L394 ?

**Comment by FabianWolff** [comment]
This is the intended behavior. Note the following detail from the [Rust reference](https://doc.rust-lang.org/reference/attributes/type_system.html#the-non_exhaustive-attribute):

> Within the defining crate, `non_exhaustive` has no effect.

Accordingly, if we move your `enum` definition into its own crate:
```rust
#[non_exhaustive]
pub enum Endless {
    JustOnce,
}
```
and access it as you have suggested:
```rust
fn main() {
    let edl = Endless::JustOnce;

    if let Endless::JustOnce = edl {
        println!("Oooh");
    } else {
        println!("Ugh");
    }

    match edl {
        Endless::JustOnce => {},
        _wildcard => {
            // in the future, something will happen here
        }
    }
}
```
it compiles without any warnings; in fact, without the wildcard, we would get
```
error[E0004]: non-exhaustive patterns: `_` not covered
```

**Comment by ohsayan** [comment]
Ah I see. Thought it to be a diagnostic error
